### ISSUE TITLE
Week8 과제 질문입니다. 

### ISSUE THREAD

**Issue by hwanginsun** [public-ai/ALAI_Homework#50]
[Exercise5_Section 7_Optimizer (1) Momentum.zip](https://github.com/public-ai/ALAI_Homework/files/3212144/Exercise5_Section.7_Optimizer.1.Momentum.zip)

1. NAG momentum을 numpy로 구현할 때, tensorflow 문법적 문제로 구현이 안되고 있습니다.

gradient를 구할 때 None값이 나와서 안되는 것 같습니다..

![image](https://user-images.githubusercontent.com/30209848/58253516-9d143800-7da3-11e9-8c93-cd681f1c6fc2.png)

![image](https://user-images.githubusercontent.com/30209848/58253540-a6050980-7da3-11e9-908a-17bd395358b0.png)

2. gradient와 momentum은 에러는 발생하지는 않지만, 이동한 거리가 정답과 다릅니다.
구현한 수식에 다른 점이 있는지 확인 부탁드립니다.
![image](https://user-images.githubusercontent.com/30209848/58253736-0eec8180-7da4-11e9-97d6-e34b13d785db.png)

**Comment by pai-plznw4me** [comment]
#1 
@hwanginsun 그 방식으로는 tensorflow 로 NAG 구현할수 없습니다. 

```
tf.gradients(y, x)
```
위 코드에서 에서 y 가 x 에 대한 함수가 아니면  gradients 는 None  이 나옵니다.
정답은 코치 쌤을 통해 공개 하겠습니다. 

#2 
요청 주신 코드에서는 아래 부분이 문제점이 있어 보입니다. 
```
        # Gradient
        loss = tf.reduce_mean(func(x, y))   
        grad_x = tf.gradients(loss, x)
        grad_y = tf.gradients(loss, y)
        
        # momentum
        before_Vt_x = tf.Variable(np.zeros_like([1], dtype=np.float32),  name='before_Vt_x')
        before_Vt_y = tf.Variable(np.zeros_like([1], dtype=np.float32),  name='before_Vt_y')
        Vt_x = tf.Variable(np.zeros_like([1], dtype=np.float32), name='Vt_x')
        Vt_y = tf.Variable(np.zeros_like([1], dtype=np.float32), name='Vt_y')
        
        
        return graph
```

```
# 위 코드는 아래와 같이 변경할 수 있을듯 합니다 .
```
        # Gradient
        loss = tf.reduce_mean(func(x, y))   
        grad_x = tf.gradients(loss, x)
        grad_y = tf.gradients(loss, y)
        
        # momentum
        vt_x = tf.Variable(0, dtype=tf.float32, name='vt_x)
        vt_y = tf.Variable(0, dtype=tf.float32, name='vt_y)
        vt_x_assign = tf.assign(vt_x, vt_x*beta - grad)
        vt_y_assign = tf.assign(vt_y, vt_y*beta - grad)
        x_assign = tf.assign_sub(x, vt_x_assign * lr)
        y_assign = tf.assign_sub(y, vt_y_assign * lr)

        step = tf.group([x_assign , y_assign])
        
        return graph
```
```

생각해야 할점은 지금 코드에서 x,y 을 따로 변수로 만들어서 사용하고 있는데 그럴 필요가 없다라는 점 입니다.
x,y 을 하나의 변수인 xy 로 묶어서 코드를 짠다면 코드가 훨씬 줄어들 것 입니다.

**Comment by hwanginsun** [comment]
정답코드로 확인해보겠습니다.
